eip/EIPS/eip-4337.md

```diff
@@ -391,78 +451,104 @@ The result `SHOULD` be set to true if and only if the request passed simulation
+  }
 }
 ```
 
-## Rationale
+### * eth_simulateUserOperation
 
-The main challenge with a purely smart contract wallet based account abstraction system is DoS safety: how can a miner including an operation make sure that it will actually pay fees, without having to first execute the entire operation? Requiring the miner to execute the entire operation opens a DoS attack vector, as an attacker could easily send many operations that pretend to pay a fee but then revert at the last moment after a long execution. Similarly, to prevent attackers from cheaply clogging the mempool, nodes in the P2P network need to check if an operation will pay a fee before they are willing to forward it.
+Simulate the validation of the user operation.
 
-In this proposal, we expect wallets to have a `validateUserOp` method that takes as input a `UserOperation`, and verify the signature and pay the fee. This method is required to be almost-pure: it is only allowed to access the storage of the wallet itself, cannot use environment opcodes (eg. `TIMESTAMP`), and can only edit the storage of the wallet, and can also send out ETH (needed to pay the entry point). The method is gas-limited by the `verificationGasLimit` of the `UserOperation`; nodes can choose to reject operations whose `verificationGasLimit` is too high. These restrictions allow miners and network nodes to simulate the verification step locally, and be confident that the result will match the result when the operation actually gets included into a block.
+**Parameters**: just like `eth_sendUserOperation`
+**Return Value**:
+* **preOpGas** gas used by validation (including preVerificationGas)
+* **prefund** amount the wallet need to pre-pay (above preOpGas and callGasLimit)
+* **deadline** if set, until what time this userOp is valid
+* **paymasterInfo** if set, current paymasterStake and paymasterUnstakeDelay
+* **aggregator** if set, the signature-aggregator used by the wallet.
 
-The entry point-based approach allows for a clean separation between verification and execution, and keeps wallets' logic simple. The alternative would be to require wallets to follow a template where they first self-call to verify and then self-call to execute (so that the execution is sandboxed and cannot cause the fee payment to revert); template-based approaches were rejected due to being harder to implement, as existing code compilation and verification tooling is not designed around template verification.
+**Error Codes:**: just like `eth_sendUserOperation`
 
-### Paymasters
+### * eth_getUserOperationTransactionByHash
 
-Paymasters facilitate transaction sponsorship, allowing third-party-designed mechanisms to pay for transactions. Many of these mechanisms _could_ be done by having the paymaster wrap a `UserOperation` with their own, but there are some important fundamental limitations to that approach:
+Return a transaction information based on a UserOperation hash (requestId) returned by `eth_sendUserOperation`
```

Comment:
TBD: should we use the term `requestId` ? it is badly named, as we don't call it "request" anywhere, and it is actually UserOpHash